src/libhirte/log/log.h

```diff
@@ -0,0 +1,146 @@
+#pragma once
+
+#include <stdarg.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <systemd/sd-journal.h>
+
+#include "libhirte/common/common.h"
+
+typedef enum LogLevel {
+        LOG_LEVEL_DEBUG,
+        LOG_LEVEL_INFO,
+        LOG_LEVEL_WARN,
+        LOG_LEVEL_ERROR,
+} LogLevel;
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+static const char *log_level_strings[] = { "DEBUG", "INFO", "WARN", "ERROR" };
+
+const char *log_level_to_string(LogLevel l) {
+        return log_level_strings[l];
+}
+
+
+typedef void (*LogFn)(
+                LogLevel lvl,
+                const char *file,
+                const char *line,
+                const char *func,
+                const char *msg,
+                const char *data);
+
+// const char *file, const char *line, const char *func,
+void hirte_log_to_journald_with_location(
+                LogLevel lvl,
+                const char *file,
+                const char *line,
+                const char *func,
+                const char *msg,
+                const char *data) {
+        // file and line are required to contain the field names and func gets its key
+        // automatically set (in contrast to what is written in the documentation).
+        // see https://www.freedesktop.org/software/systemd/man/sd_journal_print.html
+        _cleanup_free_ char *file_info = NULL;
+        _cleanup_free_ char *line_info = NULL;
+        asprintf(&file_info, "CODE_FILE=%s", file);
+        asprintf(&line_info, "CODE_LINE=%s", line);
+
+        // clang-format off
+        sd_journal_send_with_location(
+                file_info, line_info, func,
+                "HIRTE_LOG_LEVEL=%s", log_level_to_string(lvl),
+                "HIRTE_MESSAGE=%s", msg,
+                "HIRTE_DATA=%s", data,
+                NULL);
+        // clang-format on
+}
+
+void hirte_log_to_stdout_with_location(
+                LogLevel lvl,
+                const char *file,
+                const char *line,
+                const char *func,
+                const char *msg,
+                const char *data) {
+
+        time_t t = time(NULL);
+        const size_t timestamp_size = 9;
+        char timebuf[timestamp_size];
+        timebuf[strftime(timebuf, sizeof(timebuf), "%H:%M:%S", localtime(&t))] = '\0';
+
+        // clang-format off
+        fprintf(stderr, "%s %s\t%s:%s %s\t{msg: %s, data: %s}\n", 
+                timebuf, log_level_to_string(lvl), 
+                file, line, func,
+                msg, data);
+        // clang-format on
+}
+
+#define hirte_log_to_stdout hirte_log_to_stdout_with_location
+#define hirte_log_to_journald hirte_log_to_journald_with_location
+
+// NOLINTBEGIN(cppcoreguidelines-avoid-non-const-global-variables)
+static struct {
+        LogFn log_fn;
+        LogLevel level;
+        bool is_quiet;
+} HirteLogConfig;
```

Comment:
This needs to go in log.c, with the header having an extern declaration of it.

Reply:
Haven't used the `extern` keyword yet. Could you please check if I understood it correctly?

Reply:
don't think you needed it in the end